File: src/sokoenginepy/common/octoban_tessellation.py
from __future__ import annotations

from typing import Dict, Final, Tuple

from .cell_orientation import CellOrientation
from .characters import Characters
from .config import Config
from .direction import Direction
from .tessellation import index_1d, index_column, index_row, is_on_board_2d
from .tessellation_impl import PusherStepData, TessellationImpl


class OctobanTessellation(TessellationImpl):
    _LEGAL_DIRECTIONS: Final[Tuple[Direction, ...]] = (
        Direction.LEFT,
        Direction.RIGHT,
        Direction.UP,
        Direction.DOWN,
        Direction.NORTH_EAST,
        Direction.NORTH_WEST,
        Direction.SOUTH_EAST,
        Direction.SOUTH_WEST,
    )

    _CHR_TO_PUSHER_STEP: Final[Dict[str, PusherStepData]] = {
        Characters.l: PusherStepData(Direction.LEFT, False),
        Characters.L: PusherStepData(Direction.LEFT, True),
        Characters.r: PusherStepData(Direction.RIGHT, False),
        Characters.R: PusherStepData(Direction.RIGHT, True),
        Characters.u: PusherStepData(Direction.UP, False),
        Characters.U: PusherStepData(Direction.UP, True),
        Characters.d: PusherStepData(Direction.DOWN, False),
        Characters.D: PusherStepData(Direction.DOWN, True),
        Characters.w: PusherStepData(Direction.NORTH_WEST, False),
        Characters.W: PusherStepData(Direction.NORTH_WEST, True),
        Characters.e: PusherStepData(Direction.SOUTH_EAST, False),
        Characters.E: PusherStepData(Direction.SOUTH_EAST, True),
        Characters.n: PusherStepData(Direction.NORTH_EAST, False),
        Characters.N: PusherStepData(Direction.NORTH_EAST, True),
        Characters.s: PusherStepData(Direction.SOUTH_WEST, False),
        Characters.S: PusherStepData(Direction.SOUTH_WEST, True),
    }

    _PUSHER_STEP_TO_CHR: Final[Dict[PusherStepData, str]] = {
        v: k for k, v in _CHR_TO_PUSHER_STEP.items()
    }

    _NEIGHBOR_SHIFT: Final[Dict[Direction, Tuple[int, int]]] = {
        Direction.LEFT: (0, -1),
        Direction.RIGHT: (0, 1),
        Direction.UP: (-1, 0),
        Direction.DOWN: (1, 0),
        Direction.NORTH_WEST: (-1, -1),
        Direction.NORTH_EAST: (-1, 1),
        Direction.SOUTH_WEST: (1, -1),
        Direction.SOUTH_EAST: (1, 1),
    }

    def neighbor_position(
        self, position: int, direction: Direction, board_width: int, board_height: int
    ) -> int:
        if self.cell_orientation(
            position, board_width, board_height
        ) != CellOrientation.OCTAGON and direction in (
            Direction.NORTH_EAST,
            Direction.NORTH_WEST,
            Direction.SOUTH_EAST,
            Direction.SOUTH_WEST,
        ):
            return Config.NO_POS

        row = index_row(position, board_width)
        column = index_column(position, board_width)
        row_shift, column_shift = self._NEIGHBOR_SHIFT.get(direction, (None, None))

        if row_shift is None or column_shift is None:
            raise ValueError(
                f"Unsupported direction {direction} for {self.__class__.__name__}"
            )

        row += row_shift
        column += column_shift

        if is_on_board_2d(column, row, board_width, board_height):
            return index_1d(column, row, board_width)

        return Config.NO_POS

    def cell_orientation(
        self, position: int, board_width: int, board_height: int
    ) -> CellOrientation:
        if position < 0:
            raise IndexError(f"Position {position} is invalid value!")

        if board_width < 0:
            raise ValueError(f"Board width {board_width} is invalid value!")

        if board_height < 0:
            raise ValueError(f"Board height {board_height} is invalid value!")

        row = index_row(position, board_width)
        column = index_column(position, board_width)
        return (
            CellOrientation.OCTAGON
            if (column + (row % 2)) % 2 == 0
            else CellOrientation.DEFAULT
        )

File: src/libsokoengine/common/octoban_tessellation.cpp
/// @file
#include "octoban_tessellation.hpp"

#include "characters.hpp"
#include "pusher_step.hpp"

using std::invalid_argument;
using std::string;

namespace sokoengine {
namespace implementation {

static const Directions OCT_LEGAL_DIRECTIONS = {
  Direction::LEFT,
  Direction::RIGHT,
  Direction::UP,
  Direction::DOWN,
  Direction::NORTH_EAST,
  Direction::NORTH_WEST,
  Direction::SOUTH_EAST,
  Direction::SOUTH_WEST};

const Directions &OctobanTessellation::legal_directions() const {
  return OCT_LEGAL_DIRECTIONS;
}

position_t OctobanTessellation::neighbor_position(
  position_t       position,
  const Direction &direction,
  board_size_t     width,
  board_size_t     height
) const {
  position_t row = index_y(position, width), column = index_x(position, width);

  // clang-format off
  if (
    cell_orientation(position, width, height) != CellOrientation::OCTAGON
    && (
      direction == Direction::NORTH_EAST
      || direction == Direction::NORTH_WEST
      || direction == Direction::SOUTH_EAST
      || direction == Direction::SOUTH_WEST
    )
  ) {
    return Config::NO_POS;
  }
  // clang-format off

  switch (direction) {
    case Direction::LEFT:
      column += -1;
      break;
    case Direction::UP:
      row += -1;
      break;
    case Direction::RIGHT:
      column += 1;
      break;
    case Direction::DOWN:
      row += 1;
      break;
    case Direction::NORTH_WEST:
      row += -1;
      column += -1;
      break;
    case Direction::NORTH_EAST:
      row += -1;
      column += 1;
      break;
    case Direction::SOUTH_WEST:
      row += 1;
      column += -1;
      break;
    case Direction::SOUTH_EAST:
      row += 1;
      column += 1;
      break;
    default:
      throw invalid_argument(
        "Unsupported direction " + implementation::to_str(direction)
        + " for OctobanTessellation!"
      );
  }
  if (is_on_board_2d(column, row, width, height))
    return index_1d(column, row, width);
  else
    return Config::NO_POS;
}

PusherStep OctobanTessellation::char_to_pusher_step(char rv) const {
  switch (rv) {
    case Characters::l:
      return PusherStep(Direction::LEFT, Config::NO_ID);
    case Characters::L:
      return PusherStep(Direction::LEFT, Config::DEFAULT_ID);

    case Characters::u:
      return PusherStep(Direction::UP, Config::NO_ID);
    case Characters::U:
      return PusherStep(Direction::UP, Config::DEFAULT_ID);

    case Characters::r:
      return PusherStep(Direction::RIGHT, Config::NO_ID);
    case Characters::R:
      return PusherStep(Direction::RIGHT, Config::DEFAULT_ID);

    case Characters::d:
      return PusherStep(Direction::DOWN, Config::NO_ID);
    case Characters::D:
      return PusherStep(Direction::DOWN, Config::DEFAULT_ID);

    case Characters::w:
      return PusherStep(Direction::NORTH_WEST, Config::NO_ID);
    case Characters::W:
      return PusherStep(Direction::NORTH_WEST, Config::DEFAULT_ID);

    case Characters::e:
      return PusherStep(Direction::SOUTH_EAST, Config::NO_ID);
    case Characters::E:
      return PusherStep(Direction::SOUTH_EAST, Config::DEFAULT_ID);

    case Characters::n:
      return PusherStep(Direction::NORTH_EAST, Config::NO_ID);
    case Characters::N:
      return PusherStep(Direction::NORTH_EAST, Config::DEFAULT_ID);

    case Characters::s:
      return PusherStep(Direction::SOUTH_WEST, Config::NO_ID);
    case Characters::S:
      return PusherStep(Direction::SOUTH_WEST, Config::DEFAULT_ID);

    default:
      throw invalid_argument(
        string("Illegal PusherStep character '") + rv + "' in OctobanTessellation!"
      );
  }
}

char OctobanTessellation::pusher_step_to_char(const PusherStep &rv) const {
  switch (rv.direction()) {
    case Direction::LEFT:
      return rv.is_push_or_pull() ? Characters::L : Characters::l;
    case Direction::UP:
      return rv.is_push_or_pull() ? Characters::U : Characters::u;
    case Direction::RIGHT:
      return rv.is_push_or_pull() ? Characters::R : Characters::r;
    case Direction::DOWN:
      return rv.is_push_or_pull() ? Characters::D : Characters::d;
    case Direction::NORTH_WEST:
      return rv.is_push_or_pull() ? Characters::W : Characters::w;
    case Direction::SOUTH_EAST:
      return rv.is_push_or_pull() ? Characters::E : Characters::e;
    case Direction::NORTH_EAST:
      return rv.is_push_or_pull() ? Characters::N : Characters::n;
    case Direction::SOUTH_WEST:
      return rv.is_push_or_pull() ? Characters::S : Characters::s;
    default:
      throw invalid_argument(
        "Illegal PusherStep direction " + implementation::to_str(rv.direction())
        + " in OctobanTessellation!"
      );
  }
}

CellOrientation OctobanTessellation::cell_orientation(
  position_t pos, board_size_t width, board_size_t height
) const {
  position_t column = index_column(pos, width);
  position_t row    = index_row(pos, width);
  return ((column + (row % 2)) % 2 == 0) ? CellOrientation::OCTAGON
                                         : CellOrientation::DEFAULT;
}

} // namespace implementation
} // namespace sokoengine
